tests/eas/load_tracking.py

```diff
@@ -323,3 +408,163 @@ def test_task_util_avg(self, experiment, tasks):
+        error_margin = sim_mean * (ERROR_MARGIN_PCT / 100.)
+        msg = 'Saw mean value of around {}, expected {}'.format(
+            signal_stats['mean'], sim_mean)
+        self.assertAlmostEqual(sim_mean, signal_stats['mean'],
+                               delta=error_margin, msg=msg)
+
+    def _test_behaviour(self, experiment, tasks, signal_name):
+        [task] = tasks
+        signal_df = self.get_sched_task_signals(experiment, signal_name)
+        # Get instant of time when the task starts running
+        start_time = self.get_task_start_time(experiment, task)
+
+        # Get information about the task
+        phase = experiment.wload.params['profile'][task]['phases'][0]
+
+        # Create simulated PELT signal for a periodic task
+        init_pelt = self.get_initial_pelt_value(experiment, signal_df,
+                                                start_time, task)
+        peltsim, pelt_task, peltsim_df = self.get_simulated_pelt(
+            experiment, task, init_pelt
+        )
+
+        # Compare actual PELT signal with the simulated one
+        margin = 0.05
+        period_s = phase.period_ms / 1e3
+        sim_period_ms = phase.period_ms * (peltsim._sample_us / 1e6)
+        n_errors = 0
+        for entry in signal_df.iteritems():
+            trace_val = entry[1]
+            timestamp = entry[0] - start_time
+            # Next two instructions map the trace timestamp to a simulated
+            # signal timestamp. This is due to the fact that the 1 ms is
+            # actually 1024 us in the simulated signal.
+            n_periods = timestamp / period_s
+            nearest_timestamp = n_periods * sim_period_ms
+            sim_val_loc = peltsim_df.index.get_loc(nearest_timestamp,
+                                                   method='nearest')
+            sim_val = peltsim_df.pelt_value.iloc[sim_val_loc]
+            if trace_val > (sim_val * (1 + margin)) or \
+               trace_val < (sim_val * (1 - margin)):
+                self._log.error("At {} trace shows {}={}"
+                                .format(entry[0], signal_name, trace_val))
+                self._log.error("At ({}, {}) simulation shows {}={}"
+                                .format(peltsim_df.index[sim_val_loc],
+                                        signal_name,
+                                        sim_val_loc,
+                                        sim_val))
+                n_errors += 1
+
+        msg = "Total number of errors: {}/{}".format(n_errors, len(signal_df))
+        # Exclude possible outliers (these may be due to a kernel thread that
+        # for some reason gets coscheduled with our workload).
+        self.assertLess(n_errors/len(signal_df), margin, msg)
+
+    @experiment_test
+    def test_util_avg_range(self, experiment, tasks):
+        """
+        Test PELT util_avg stable range for a 50% periodic task
+        """
+        return self._test_range(experiment, tasks, 'util_avg')
```

Comment:
Are the `_test_range` and `_test_behavoiur` helpers actually needed? Looks like you can just put the code directly in the test method.

Reply:
They can be used for testing `load_avg` as well, that's why I kept those separate.